test/test.oauth2.ts

```diff
@@ -1338,15 +1378,95 @@ describe('oauth2', () => {
+    it('should call refreshHandler in getRequestHeaders() on token expiration and refreshHandler is available', async () => {
+      const expectedRefreshedAccessToken = {
+        access_token: 'access_token',
+        expiry_date: new Date().getTime() + 3600 * 1000,
+      };
+      client.refreshHandler = async () => {
+        return expectedRefreshedAccessToken;
+      };
       client.setCredentials({
         access_token: 'initial-access-token',
         expiry_date: new Date().getTime() - 1000,
       });
+      const expectedMetadata = {
+        Authorization: 'Bearer access_token',
+      };
+
+      const requestMetaData = await client.getRequestHeaders(
+        'http://example.com'
+      );
+
+      assert.deepStrictEqual(requestMetaData, expectedMetadata);
+    });
+
+    it('should return authorization header on current access token if it has not expired', async () => {
+      client.setCredentials({
+        access_token: 'initial-access-token',
+        expiry_date: new Date().getTime() + 3600 * 1000,
+      });
+      const expectedMetadata = {
+        Authorization: 'Bearer initial-access-token',
+      };
+
+      const requestMetaData = await client.getRequestHeaders(
+        'http://example.com'
+      );
+
+      assert.deepStrictEqual(requestMetaData, expectedMetadata);
+    });
+
+    it('should throw if tries to refresh but no refreshHandler callback or refresh token is available', async () => {
+      client.setCredentials({
+        access_token: 'initial-access-token',
+        expiry_date: new Date().getTime() - 1000,
+      });
+
       await assert.rejects(
         client.getRequestHeaders('http://example.com'),
         /No refresh token is set./
       );
     });
+
+    it('should call refreshHandler in getAccessToken() on expiration and no refresh token available', async () => {
+      const expectedRefreshedAccessToken = {
+        access_token: 'access_token',
+        expiry_date: new Date().getTime() + 3600 * 1000,
+      };
+      client.refreshHandler = async () => {
+        return expectedRefreshedAccessToken;
+      };
+      client.setCredentials({
+        access_token: 'initial-access-token',
+        expiry_date: new Date().getTime() - 1000,
+      });
+
+      const refreshedAccessToken = await client.getAccessToken();
+
+      assert.strictEqual(
+        refreshedAccessToken.token,
+        expectedRefreshedAccessToken.access_token
+      );
+    });
+
+    it('should throw error if refreshHandler callback response is missing an access token', async () => {
+      const expectedRefreshedAccessToken = {
+        access_token: '',
+        expiry_date: new Date().getTime() + 3600 * 1000,
+      };
+      client.refreshHandler = async () => {
+        return expectedRefreshedAccessToken;
+      };
+      client.setCredentials({
+        access_token: 'initial-access-token',
+        expiry_date: new Date().getTime() - 1000,
+      });
```

Comment:
Please minimize using `setCredentials()` to inject expired credentials unless you are specifically testing expiration flows. Please see my reasoning in the previous PRs I reviewed. The oauth2client should be able to be initialized with access tokens and solely with a `refreshHandler`.

```
const client = new OAuth2Client();
client.refreshHandler = async () => {...};
```